// react-instantsearch/dist/cjs/ui/Snippet.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "Snippet", {
    enumerable: true,
    get: function() {
        return Snippet;
    }
});
var _interop_require_default = require("@swc/helpers/_/_interop_require_default");
var _object_spread = require("@swc/helpers/_/_object_spread");
var _object_without_properties = require("@swc/helpers/_/_object_without_properties");
var _instantsearchuicomponents = require("instantsearch-ui-components");
var _react = /*#__PURE__*/ _interop_require_default._(require("react"));
var _InternalHighlight = require("./InternalHighlight");
function Snippet(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, props = _object_without_properties._(_0, [
        "classNames"
    ]);
    return /*#__PURE__*/ _react.default.createElement(_InternalHighlight.InternalHighlight, _object_spread._({
        classNames: {
            root: (0, _instantsearchuicomponents.cx)('ais-Snippet', classNames.root),
            highlighted: (0, _instantsearchuicomponents.cx)('ais-Snippet-highlighted', classNames.highlighted),
            nonHighlighted: (0, _instantsearchuicomponents.cx)('ais-Snippet-nonHighlighted', classNames.nonHighlighted),
            separator: (0, _instantsearchuicomponents.cx)('ais-Snippet-separator', classNames.separator)
        }
    }, props));
}
