// react-instantsearch/dist/es/widgets/Chat.js
import { _ } from '@swc/helpers/esm/_define_property.js';
import { _ as _$3 } from '@swc/helpers/esm/_object_spread.js';
import { _ as _$4 } from '@swc/helpers/esm/_object_spread_props.js';
import { _ as _$1 } from '@swc/helpers/esm/_object_without_properties.js';
import { _ as _$2 } from '@swc/helpers/esm/_sliced_to_array.js';
import { createChatComponent } from 'instantsearch-ui-components';
import { SearchIndexToolType, RecommendToolType, MemorizeToolType, MemorySearchToolType, PonderToolType } from 'instantsearch.js/es/lib/chat/index.js';
export { MemorizeToolType, MemorySearchToolType, PonderToolType, RecommendToolType, SearchIndexToolType } from 'instantsearch.js/es/lib/chat/index.js';
import React, { createElement, Fragment, useState, useRef, useMemo, useEffect } from 'react';
import { useInstantSearch, useChat } from 'react-instantsearch-core';
import { useStickToBottom } from '../lib/useStickToBottom.js';
import { createCarouselTool } from './chat/tools/SearchIndexTool.js';

var ChatUiComponent = createChatComponent({
    createElement: createElement,
    Fragment: Fragment
});
function createDefaultTools(itemComponent, getSearchPageURL) {
    var _obj;
    return _obj = {}, _(_obj, SearchIndexToolType, createCarouselTool(true, itemComponent, getSearchPageURL)), _(_obj, RecommendToolType, createCarouselTool(false, itemComponent, getSearchPageURL)), _(_obj, MemorizeToolType, {}), _(_obj, MemorySearchToolType, {}), _(_obj, PonderToolType, {}), _obj;
}
function Chat(_0) {
    var userTools = _0.tools, toggleButtonProps = _0.toggleButtonProps, headerProps = _0.headerProps, messagesProps = _0.messagesProps, promptProps = _0.promptProps, itemComponent = _0.itemComponent, toggleButtonComponent = _0.toggleButtonComponent, toggleButtonIconComponent = _0.toggleButtonIconComponent, headerComponent = _0.headerComponent, headerTitleIconComponent = _0.headerTitleIconComponent, headerCloseIconComponent = _0.headerCloseIconComponent, headerMinimizeIconComponent = _0.headerMinimizeIconComponent, headerMaximizeIconComponent = _0.headerMaximizeIconComponent, messagesLoaderComponent = _0.messagesLoaderComponent, messagesErrorComponent = _0.messagesErrorComponent, promptComponent = _0.promptComponent, promptHeaderComponent = _0.promptHeaderComponent, promptFooterComponent = _0.promptFooterComponent, assistantMessageLeadingComponent = _0.assistantMessageLeadingComponent, assistantMessageFooterComponent = _0.assistantMessageFooterComponent, userMessageLeadingComponent = _0.userMessageLeadingComponent, userMessageFooterComponent = _0.userMessageFooterComponent, actionsComponent = _0.actionsComponent, suggestionsComponent = _0.suggestionsComponent, classNames = _0.classNames, _0_translations = _0.translations, translations = _0_translations === void 0 ? {} : _0_translations, title = _0.title, getSearchPageURL = _0.getSearchPageURL, props = _$1(_0, [
        "tools",
        "toggleButtonProps",
        "headerProps",
        "messagesProps",
        "promptProps",
        "itemComponent",
        "toggleButtonComponent",
        "toggleButtonIconComponent",
        "headerComponent",
        "headerTitleIconComponent",
        "headerCloseIconComponent",
        "headerMinimizeIconComponent",
        "headerMaximizeIconComponent",
        "messagesLoaderComponent",
        "messagesErrorComponent",
        "promptComponent",
        "promptHeaderComponent",
        "promptFooterComponent",
        "assistantMessageLeadingComponent",
        "assistantMessageFooterComponent",
        "userMessageLeadingComponent",
        "userMessageFooterComponent",
        "actionsComponent",
        "suggestionsComponent",
        "classNames",
        "translations",
        "title",
        "getSearchPageURL"
    ]);
    var promptTranslations = translations.prompt, headerTranslations = translations.header, messageTranslations = translations.message, messagesTranslations = translations.messages;
    var _useInstantSearch = useInstantSearch(), indexUiState = _useInstantSearch.indexUiState, setIndexUiState = _useInstantSearch.setIndexUiState;
    var _useState = _$2(useState(false), 2), maximized = _useState[0], setMaximized = _useState[1];
    var promptRef = useRef(null);
    var _useStickToBottom = useStickToBottom({
        initial: 'smooth',
        resize: 'smooth'
    }), scrollRef = _useStickToBottom.scrollRef, contentRef = _useStickToBottom.contentRef, scrollToBottom = _useStickToBottom.scrollToBottom, isAtBottom = _useStickToBottom.isAtBottom;
    var tools = useMemo(function() {
        var defaults = createDefaultTools(itemComponent, getSearchPageURL);
        return _$3({}, defaults, userTools);
    }, [
        getSearchPageURL,
        itemComponent,
        userTools
    ]);
    var chatState = useChat(_$4(_$3({}, props), {
        tools: tools
    }));
    var messages = chatState.messages, sendMessage = chatState.sendMessage, status = chatState.status, regenerate = chatState.regenerate, stop = chatState.stop; chatState.error; var input = chatState.input, setInput = chatState.setInput, open = chatState.open, setOpen = chatState.setOpen, isClearing = chatState.isClearing, clearMessages = chatState.clearMessages, onClearTransitionEnd = chatState.onClearTransitionEnd, toolsFromConnector = chatState.tools, suggestions = chatState.suggestions;
    var wasOpenRef = useRef(false);
    useEffect(function() {
        var shouldFocusPrompt = !wasOpenRef.current && open;
        if (shouldFocusPrompt) {
            window.requestAnimationFrame(function() {
                var _promptRef_current;
                (_promptRef_current = promptRef.current) === null || _promptRef_current === void 0 ? void 0 : _promptRef_current.focus();
            });
        }
        wasOpenRef.current = open;
    }, [
        open
    ]);
    return /*#__PURE__*/ React.createElement(ChatUiComponent, {
        title: title,
        open: open,
        maximized: maximized,
        headerComponent: headerComponent,
        promptComponent: promptComponent,
        toggleButtonComponent: toggleButtonComponent,
        suggestionsComponent: suggestionsComponent,
        toggleButtonProps: _$3({
            open: open,
            onClick: function onClick() {
                return setOpen(!open);
            },
            toggleIconComponent: toggleButtonIconComponent
        }, toggleButtonProps),
        headerProps: _$3({
            onClose: function onClose() {
                return setOpen(false);
            },
            maximized: maximized,
            onToggleMaximize: function onToggleMaximize() {
                return setMaximized(!maximized);
            },
            onClear: clearMessages,
            canClear: Boolean(messages === null || messages === void 0 ? void 0 : messages.length) && !isClearing,
            titleIconComponent: headerTitleIconComponent,
            closeIconComponent: headerCloseIconComponent,
            minimizeIconComponent: headerMinimizeIconComponent,
            maximizeIconComponent: headerMaximizeIconComponent,
            translations: headerTranslations
        }, headerProps),
        messagesProps: _$3({
            status: status,
            onReload: function onReload(messageId) {
                return regenerate({
                    messageId: messageId
                });
            },
            onClose: function onClose() {
                return setOpen(false);
            },
            messages: messages,
            tools: toolsFromConnector,
            indexUiState: indexUiState,
            setIndexUiState: setIndexUiState,
            isClearing: isClearing,
            onClearTransitionEnd: onClearTransitionEnd,
            isScrollAtBottom: isAtBottom,
            scrollRef: scrollRef,
            contentRef: contentRef,
            onScrollToBottom: scrollToBottom,
            loaderComponent: messagesLoaderComponent,
            errorComponent: messagesErrorComponent,
            actionsComponent: actionsComponent,
            assistantMessageProps: _$3({
                leadingComponent: assistantMessageLeadingComponent,
                footerComponent: assistantMessageFooterComponent
            }, messagesProps === null || messagesProps === void 0 ? void 0 : messagesProps.assistantMessageProps),
            userMessageProps: _$3({
                leadingComponent: userMessageLeadingComponent,
                footerComponent: userMessageFooterComponent
            }, messagesProps === null || messagesProps === void 0 ? void 0 : messagesProps.userMessageProps),
            translations: messagesTranslations,
            messageTranslations: messageTranslations
        }, messagesProps),
        promptProps: _$3({
            promptRef: promptRef,
            status: status,
            value: input,
            translations: promptTranslations,
            onInput: function onInput(event) {
                setInput(event.currentTarget.value);
            },
            onSubmit: function onSubmit() {
                sendMessage({
                    text: input
                });
                setInput('');
            },
            onStop: function onStop() {
                stop();
            },
            headerComponent: promptHeaderComponent,
            footerComponent: promptFooterComponent
        }, promptProps),
        suggestionsProps: {
            suggestions: suggestions,
            onSuggestionClick: function onSuggestionClick(suggestion) {
                sendMessage({
                    text: suggestion
                });
            }
        },
        classNames: classNames
    });
}

export { Chat, createDefaultTools };
