// instantsearch.js/es/lib/utils/index.js
export { addWidgetId, resetWidgetId } from './addWidgetId.js';
export { capitalize } from './capitalize.js';
export { checkIndexUiState } from './checkIndexUiState.js';
export { checkRendering } from './checkRendering.js';
export { clearRefinements } from './clearRefinements.js';
export { concatHighlightedParts } from './concatHighlightedParts.js';
export { createConcurrentSafePromise } from './createConcurrentSafePromise.js';
export { createSendEventForFacet } from './createSendEventForFacet.js';
export { _buildEventPayloadsForHits, createBindEventForHits, createSendEventForHits } from './createSendEventForHits.js';
export { setIndexHelperState } from './setIndexHelperState.js';
export { isIndexWidget } from './isIndexWidget.js';
export { debounce } from './debounce.js';
export { defer } from './defer.js';
export { createDocumentationLink, createDocumentationMessageGenerator } from './documentation.js';
export { TAG_PLACEHOLDER, TAG_REPLACEMENT, escapeFacets, escapeHits } from './escape-highlight.js';
export { escape, unescape } from './escape-html.js';
export { escapeFacetValue, unescapeFacetValue } from './escapeFacetValue.js';
export { find } from './find.js';
export { findIndex } from './findIndex.js';
export { aroundLatLngToPosition, insideBoundingBoxToBoundingBox } from './geo-search.js';
export { getAlgoliaAgent } from './getAlgoliaAgent.js';
export { getAppIdAndApiKey } from './getAppIdAndApiKey.js';
export { getContainerNode } from './getContainerNode.js';
export { getHighlightedParts } from './getHighlightedParts.js';
export { getHighlightFromSiblings } from './getHighlightFromSiblings.js';
export { getObjectType } from './getObjectType.js';
export { getPropertyByPath } from './getPropertyByPath.js';
export { getRefinements } from './getRefinements.js';
export { getWidgetAttribute } from './getWidgetAttribute.js';
export { addAbsolutePosition } from './hits-absolute-position.js';
export { addQueryID } from './hits-query-id.js';
export { hydrateRecommendCache } from './hydrateRecommendCache.js';
export { hydrateSearchClient } from './hydrateSearchClient.js';
export { isDomElement } from './isDomElement.js';
export { isEqual } from './isEqual.js';
export { isFacetRefined } from './isFacetRefined.js';
export { isFiniteNumber } from './isFiniteNumber.js';
export { isPlainObject } from './isPlainObject.js';
export { isSpecialClick } from './isSpecialClick.js';
export { walkIndex } from './walkIndex.js';
export { deprecate, warn, warning } from './logger.js';
export { mergeSearchParameters } from './mergeSearchParameters.js';
export { omit } from './omit.js';
export { noop } from './noop.js';
export { range } from './range.js';
export { createInitArgs, createRenderArgs } from './render-args.js';
export { resolveSearchParameters } from './resolveSearchParameters.js';
export { reverseHighlightedParts } from './reverseHighlightedParts.js';
export { safelyRunOnBrowser } from './safelyRunOnBrowser.js';
export { deserializePayload, serializePayload } from './serializer.js';
export { toArray } from './toArray.js';
export { uniq } from './uniq.js';
