// react-instantsearch/dist/es/ui/ToggleRefinement.js
import { _ as _$2 } from '@swc/helpers/esm/_object_spread.js';
import { _ as _$1 } from '@swc/helpers/esm/_object_spread_props.js';
import { _ } from '@swc/helpers/esm/_object_without_properties.js';
import { cx } from 'instantsearch-ui-components';
import React from 'react';

function ToggleRefinement(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, checked = _0.checked, onChange = _0.onChange, label = _0.label, props = _(_0, [
        "classNames",
        "checked",
        "onChange",
        "label"
    ]);
    return /*#__PURE__*/ React.createElement("div", _$1(_$2({}, props), {
        className: cx('ais-ToggleRefinement', classNames.root, props.className)
    }), /*#__PURE__*/ React.createElement("label", {
        className: cx('ais-ToggleRefinement-label', classNames.label)
    }, /*#__PURE__*/ React.createElement("input", {
        className: cx('ais-ToggleRefinement-checkbox', classNames.checkbox),
        type: "checkbox",
        checked: checked,
        onChange: function onChange1(event) {
            onChange(event.target.checked);
        }
    }), /*#__PURE__*/ React.createElement("span", {
        className: cx('ais-ToggleRefinement-labelText', classNames.labelText)
    }, label)));
}

export { ToggleRefinement };
