// react-instantsearch/dist/es/widgets/index.umd.js
export { EXPERIMENTAL_Autocomplete } from './Autocomplete.js';
export { Breadcrumb } from './Breadcrumb.js';
export { FilterSuggestions } from './FilterSuggestions.js';
export { ClearRefinements } from './ClearRefinements.js';
export { CurrentRefinements } from './CurrentRefinements.js';
export { FrequentlyBoughtTogether } from './FrequentlyBoughtTogether.js';
export { HierarchicalMenu } from './HierarchicalMenu.js';
export { Highlight } from './Highlight.js';
export { Hits } from './Hits.js';
export { HitsPerPage } from './HitsPerPage.js';
export { InfiniteHits } from './InfiniteHits.js';
export { LookingSimilar } from './LookingSimilar.js';
export { Menu } from './Menu.js';
export { Pagination } from './Pagination.js';
export { PoweredBy } from './PoweredBy.js';
export { RangeInput } from './RangeInput.js';
export { RefinementList } from './RefinementList.js';
export { RelatedProducts } from './RelatedProducts.js';
export { ReverseHighlight } from './ReverseHighlight.js';
export { SearchBox } from './SearchBox.js';
export { Snippet } from './Snippet.js';
export { SortBy } from './SortBy.js';
export { Stats } from './Stats.js';
export { ToggleRefinement } from './ToggleRefinement.js';
export { TrendingItems } from './TrendingItems.js';

var Chat = function Chat() {
    throw new Error('"<Chat>" is not available from the UMD build.\n\nPlease use React InstantSearch with a packaging system:\nhttps://www.algolia.com/doc/guides/building-search-ui/installation/react/#install-react-instantsearch-as-an-npm-package');
};
var SearchIndexToolType = undefined;
var RecommendToolType = undefined;
var MemorizeToolType = undefined;
var MemorySearchToolType = undefined;
var PonderToolType = undefined;
var createDefaultTools = function createDefaultTools() {};

export { Chat, MemorizeToolType, MemorySearchToolType, PonderToolType, RecommendToolType, SearchIndexToolType, createDefaultTools };
