// react-instantsearch/dist/cjs/widgets/index.umd.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get Chat () {
        return Chat;
    },
    get MemorizeToolType () {
        return MemorizeToolType;
    },
    get MemorySearchToolType () {
        return MemorySearchToolType;
    },
    get PonderToolType () {
        return PonderToolType;
    },
    get RecommendToolType () {
        return RecommendToolType;
    },
    get SearchIndexToolType () {
        return SearchIndexToolType;
    },
    get createDefaultTools () {
        return createDefaultTools;
    }
});
var _export_star = require("@swc/helpers/_/_export_star");
_export_star._(require("./Autocomplete"), exports);
_export_star._(require("./Breadcrumb"), exports);
_export_star._(require("./FilterSuggestions"), exports);
_export_star._(require("./ClearRefinements"), exports);
_export_star._(require("./CurrentRefinements"), exports);
_export_star._(require("./FrequentlyBoughtTogether"), exports);
_export_star._(require("./HierarchicalMenu"), exports);
_export_star._(require("./Highlight"), exports);
_export_star._(require("./Hits"), exports);
_export_star._(require("./HitsPerPage"), exports);
_export_star._(require("./InfiniteHits"), exports);
_export_star._(require("./LookingSimilar"), exports);
_export_star._(require("./Menu"), exports);
_export_star._(require("./Pagination"), exports);
_export_star._(require("./PoweredBy"), exports);
_export_star._(require("./RangeInput"), exports);
_export_star._(require("./RefinementList"), exports);
_export_star._(require("./RelatedProducts"), exports);
_export_star._(require("./ReverseHighlight"), exports);
_export_star._(require("./SearchBox"), exports);
_export_star._(require("./Snippet"), exports);
_export_star._(require("./SortBy"), exports);
_export_star._(require("./Stats"), exports);
_export_star._(require("./ToggleRefinement"), exports);
_export_star._(require("./TrendingItems"), exports);
var Chat = function Chat() {
    throw new Error('"<Chat>" is not available from the UMD build.\n\nPlease use React InstantSearch with a packaging system:\nhttps://www.algolia.com/doc/guides/building-search-ui/installation/react/#install-react-instantsearch-as-an-npm-package');
};
var SearchIndexToolType = undefined;
var RecommendToolType = undefined;
var MemorizeToolType = undefined;
var MemorySearchToolType = undefined;
var PonderToolType = undefined;
var createDefaultTools = function createDefaultTools() {};
