// instantsearch.js/cjs/components/ReverseHighlight/ReverseHighlight.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "ReverseHighlight", {
    enumerable: true,
    get: function() {
        return ReverseHighlight;
    }
});
var _object_spread = require("@swc/helpers/_/_object_spread");
var _object_without_properties = require("@swc/helpers/_/_object_without_properties");
var _instantsearchuicomponents = require("instantsearch-ui-components");
var _preact = require("preact");
var _InternalHighlight = require("../InternalHighlight/InternalHighlight");
function ReverseHighlight(_0) {
    var _0_classNames = _0.classNames, classNames = _0_classNames === void 0 ? {} : _0_classNames, props = _object_without_properties._(_0, [
        "classNames"
    ]);
    return /*#__PURE__*/ (0, _preact.h)(_InternalHighlight.InternalHighlight, _object_spread._({
        classNames: {
            root: (0, _instantsearchuicomponents.cx)('ais-ReverseHighlight', classNames.root),
            highlighted: (0, _instantsearchuicomponents.cx)('ais-ReverseHighlight-highlighted', classNames.highlighted),
            nonHighlighted: (0, _instantsearchuicomponents.cx)('ais-ReverseHighlight-nonHighlighted', classNames.nonHighlighted),
            separator: (0, _instantsearchuicomponents.cx)('ais-ReverseHighlight-separator', classNames.separator)
        }
    }, props));
}
